src/server/game-server.ts

```diff
@@ -1,62 +1,85 @@
+
+import { SEND_STATE_UPDATE, StatePayload } from './server-interface';
+import { SEND_FRAMES } from '../client/client-interface';
 
 const TICKS_PER_SECOND = 60;
+const DEV_DELAY = 0; // Delays the updates sent to clients to simulate slower connections
 export class GameServer {
   game: GameState;
   sockets: SocketIO.Socket[];
-  frameBuffer: InputFrame[];
   disconnects: string[];
   lastTick: number;
   gameStartPromise: Promise<{}>;
-  startGame: () => void; // resolves the promise
+
+  playerSockets: Map<string, SocketIO.Socket>;
+  lastUpdateTimestamps: Map<SocketIO.Socket, number>; //
+  frameBuffers: {[id:string]: InputFrame[]};
 
   constructor(sockets: SocketIO.Socket[]) {
     this.game = Game.init();
-    this.frameBuffer = [];
+    this.frameBuffers = {};
     this.disconnects = [];
+    this.lastUpdateTimestamps = new Map();
+    this.playerSockets = new Map();
     this.lastTick = Date.now();
     this.sockets = sockets;
     for (var socket of this.sockets) {
       let player = Game.addPlayer(this.game);
-      socket.on('new frames', this.acceptFrames.bind(this));
+      socket.on(SEND_FRAMES, (frames: InputFrame[]) => this.acceptFrames(frames, player.id));
       socket.on('disconnect', () => this.onDisconnection(player.id));
       socket.emit('start game', {
         playerId: player.id,
         gameState: this.game,
       });
+      this.playerSockets.set(player.id, socket);
     }
     this.tick();
   }
 
   tick() {
-    let {game, frameBuffer, disconnects, lastTick} = this;
-    this.frameBuffer = [];
-    this.disconnects = [];
+    let {game, frameBuffers, disconnects, lastTick} = this;
     let delta = Date.now() - lastTick;
     this.lastTick += delta;
 
     // Handle client disconnect
     disconnects.forEach(id => Game.removePlayer(game, id));
-    let events = Game.applyInputs(game, frameBuffer);
+
+    // Update the "most recent update" timestamp for each player to
+    // inform it of how delayed its server payload is
+    _.forIn(frameBuffers, (frames, playerId) =>
+      this.lastUpdateTimestamps.set(this.playerSockets.get(playerId), _.last(frames).timestamp));
+
+    let frames = _.flatten(_.values(frameBuffers));
+    let events = Game.applyInputs(game, frames);
     events = events.concat(Game.update(game, delta));
     Game.resolveEvents(game, events);
     this.sendState();
+    this.frameBuffers = {};
+    this.disconnects = [];
     setTimeout(this.tick.bind(this), (1/TICKS_PER_SECOND) * 1000);
   }
 
   onDisconnection(playerId: string) {
     this.disconnects.push(playerId);
   }
 
-  acceptFrames(frames: InputFrame[]) {
-    this.frameBuffer = this.frameBuffer.concat(frames);
+  acceptFrames(frames: InputFrame[], playerId: string) {
+    this.frameBuffers[playerId] = (this.frameBuffers[playerId] || []).concat(frames);
   }
 
   sendState() {
-    for (var socket of this.sockets) {
-      socket.emit('state update', this.game);
-    }
+      for (var socket of this.sockets) {
```

Comment:
Perhaps this whole thing we can do a big if statement, basically don't do extra work if DEV_DELAY = 0?